crates/services/p2p/src/discovery.rs

```diff
@@ -187,159 +233,102 @@ impl NetworkBehaviour for DiscoveryBehaviour {
-    fn build_fuel_discovery(
-        bootstrap_nodes: Vec<Multiaddr>,
-    ) -> (Swarm<DiscoveryBehaviour>, Multiaddr, PeerId) {
-        let keypair = Keypair::generate_secp256k1();
-        let public_key = keypair.public();
+    use libp2p_swarm_test::SwarmExt;
+    use std::sync::Arc;
 
-        let noise_keys = noise::Keypair::<noise::X25519Spec>::new()
-            .into_authentic(&keypair)
-            .unwrap();
-
-        let transport = core::transport::MemoryTransport::new()
-            .upgrade(core::upgrade::Version::V1)
-            .authenticate(noise::NoiseConfig::xx(noise_keys).into_authenticated())
-            .multiplex(yamux::YamuxConfig::default())
-            .boxed();
-
-        let behaviour = {
+    fn build_behavior_fn(
+        bootstrap_nodes: Vec<Multiaddr>,
+    ) -> impl FnOnce(Keypair) -> DiscoveryBehaviour {
+        |keypair| {
             let mut config = DiscoveryConfig::new(
                 keypair.public().to_peer_id(),
                 "test_network".into(),
             );
             config
-                .discovery_limit(50)
+                .max_peers_connected(50)
                 .with_bootstrap_nodes(bootstrap_nodes)
-                .set_connection_idle_timeout(Duration::from_secs(120))
-                .with_random_walk(Duration::from_secs(5));
+                .with_random_walk(Duration::from_millis(500));
 
             config.finish()
-        };
+        }
+    }
+
+    /// helper function for building Discovery Behaviour for testing
+    fn build_fuel_discovery(
+        bootstrap_nodes: Vec<Multiaddr>,
+    ) -> (Swarm<DiscoveryBehaviour>, Multiaddr, PeerId) {
+        let behaviour_fn = build_behavior_fn(bootstrap_nodes);
 
         let listen_addr: Multiaddr = Protocol::Memory(rand::random::<u64>()).into();
-        let swarm_builder = SwarmBuilder::without_executor(
-            transport,
-            behaviour,
-            keypair.public().to_peer_id(),
-        )
-        .dial_concurrency_factor(NonZeroU8::new(1).expect("1 > 0"));
 
-        let mut swarm = swarm_builder.build();
+        let mut swarm = Swarm::new_ephemeral(behaviour_fn);
 
         swarm
             .listen_on(listen_addr.clone())
             .expect("swarm should start listening");
 
-        (swarm, listen_addr, PeerId::from_public_key(&public_key))
+        let peer_id = swarm.local_peer_id().to_owned();
+
+        (swarm, listen_addr, peer_id)
     }
 
     // builds 25 discovery swarms,
     // initially, only connects first_swarm to the rest of the swarms
     // after that each swarm uses kademlia to discover other swarms
     // test completes after all swarms have connected to each other
+    // TODO: This used to fail with any connection closures, but that was causing a lot of failed tests.
+    //   Now it allows for many connection closures before failing. We don't know what caused the
+    //   connections to start failing, but had something to do with upgrading `libp2p`.
```

Comment:
Do we have a ticket to keep track of this issue?

Reply:
Not yet. I can create one for investigation. I mostly left this as a note to myself and reviewers to have a discussion about the change in behavior.

Reply:
No. I've tried re-running since we updated the codec and everything. It is still flaky.

Reply:
https://github.com/FuelLabs/fuel-core/issues/1562